sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala

```diff
@@ -0,0 +1,445 @@
+      val thread = new Thread() { override def run() = { f } }
+      thread.setDaemon(daemon)
+      thread.start()
+      thread
+    }
+  }
+
+  def apply(command: Seq[String]): Stream[String] = {
+    val streamed = BlockingStreamed[String](true)
+    val process = command.run(BasicIO(false, streamed.process, None))
+    Spawn(streamed.done(process.exitValue()))
+    streamed.stream()
+  }
+}
+
+trait DataGenerator extends Serializable {
+  protected val toolsDir: String
+
+  protected def generateData(
+      sparkContext: SparkContext,
+      tableName: String,
+      partitions: Int,
+      scaleFactor: Int,
+      command: String): RDD[String] = {
+    val generatedData = {
+      sparkContext.parallelize(1 to partitions, partitions).flatMap { i =>
+        val localToolsDir = if (new java.io.File(toolsDir).exists) {
+          toolsDir
+        } else if (new java.io.File(s"/$toolsDir").exists) {
+          s"/$toolsDir"
+        } else {
+          throw new IllegalStateException(
+            s"Could not find tools at $toolsDir or /$toolsDir. Run install")
+        }
+
+        val parallel = if (partitions > 1) s"-parallel $partitions -child $i" else ""
+        val commands = Seq("bash", "-c", s"cd $localToolsDir && $command $parallel")
+        BlockingLineStream(commands)
+      }
+    }
+
+    generatedData.setName(s"$tableName, sf=$scaleFactor, strings")
+    generatedData
+  }
+
+  def generate(
+      sparkContext: SparkContext,
+      tableName: String,
+      partitions: Int,
+      scaleFactor: Int): RDD[String]
+}
+
+class Dsdgen(dsdgenDir: String) extends DataGenerator with Serializable {
+  override protected val toolsDir: String = s"$dsdgenDir"
+
+  override def generate(
+      sparkContext: SparkContext,
+      tableName: String,
+      partitions: Int,
+      scaleFactor: Int): RDD[String] = {
+    val command = s"./dsdgen -table $tableName -filter Y -scale $scaleFactor -RNGSEED 100"
+    generateData(sparkContext, tableName, partitions, scaleFactor, command)
+  }
+}
+
+class Dbgen(dbgenDir: String, params: Seq[String])
+  extends DataGenerator with Serializable {
+  override protected val toolsDir: String = s"$dbgenDir"
+  override def generate(
+      sparkContext: SparkContext,
+      tableName: String,
+      partitions: Int,
+      scaleFactor: Int): RDD[String] = {
+    val shortTableNames = Map(
+      "customer" -> "c",
+      "lineitem" -> "L",
+      "nation" -> "n",
+      "orders" -> "O",
+      "part" -> "P",
+      "region" -> "r",
+      "supplier" -> "s",
+      "partsupp" -> "S"
+    )
+    val paramsString = params.mkString(" ")
+    val command = s"./dbgen -q $paramsString -T ${shortTableNames(tableName)} -s $scaleFactor"
+    generateData(sparkContext, tableName, partitions, scaleFactor, command)
+  }
+}
+
+trait TableGenerator extends Serializable with Logging {
+  protected val dataGenerator: DataGenerator
+  protected val sparkSQLContext: SQLContext
```

Comment:
```suggestion
  protected val sqlContext: SQLContext
```

Reply:
BTW. SQLContext is an old API. It seems it's the time use `SparkSession` instead. cc @cloud-fan @MaxGekk @wangyum

Reply:
The sqlContext is extracted out of the same SparkSession, I don't think this is required, even if we pass it, in the genData, we will have to again get hold of the sqlContext or catalog to operate on table. Let me know if this is absolutely required. Thanks!

Reply:
I created https://github.com/apache/spark/pull/41900 to replace `SQLContext` with `SparkSession` for `GenTPCDSData` and it works good.

@surnaik Could you wait for https://github.com/apache/spark/pull/41900 merged and update this PR ?

Reply:
Sounds good. Thanks, please update once #41900 is merged.